tools/c7n_azure/c7n_azure/container_host/host.py

```diff
@@ -0,0 +1,415 @@
+# Copyright 2019 Microsoft Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import base64
+import json
+import logging
+import os
+import tempfile
+from datetime import datetime
+
+import yaml
+from apscheduler.schedulers.blocking import BlockingScheduler
+from apscheduler.triggers.cron import CronTrigger
+from azure.common import AzureHttpError
+from azure.mgmt.eventgrid.models import \
+    StorageQueueEventSubscriptionDestination, StringInAdvancedFilter, EventSubscriptionFilter
+from c7n_azure import entry, constants
+from c7n_azure.azure_events import AzureEventSubscription, AzureEvents
+from c7n_azure.provider import Azure
+from c7n_azure.session import Session
+from c7n_azure.storage_utils import StorageUtilities as Storage
+from c7n_azure.utils import ResourceIdParser
+
+from c7n.config import Config
+from c7n.policy import PolicyCollection
+from c7n.resources import load_resources
+from c7n.utils import local_session
+
+log = logging.getLogger("c7n_azure.container-host")
+max_dequeue_count = 2
+policy_update_seconds = 60
+queue_poll_seconds = 15
+jitter_seconds = 10
+queue_timeout_seconds = 5 * 60
+queue_message_count = 5
+
+
+class Host:
+
+    def __init__(self):
+        logging.basicConfig(level=logging.INFO, format='%(message)s')
+        log.info("Running Azure Cloud Custodian Self-Host")
+
+        if not Host.has_required_params():
+            return
+
+        load_resources()
+        self.session = local_session(Session)
+
+        # Load configuration
+        self.options = Host.build_options()
+        self.policy_storage_uri = os.getenv(constants.ENV_CONTAINER_POLICY_STORAGE)
+        self.event_queue_name = os.getenv(constants.ENV_CONTAINER_EVENT_QUEUE_NAME)
+        self.event_queue_id = os.getenv(constants.ENV_CONTAINER_EVENT_QUEUE_ID)
+
+        # Prepare storage bits
+        self.policy_blob_client = None
+        self.blob_cache = {}
+        self.queue_storage_account = self.prepare_queue_storage(
+            self.event_queue_id,
+            self.event_queue_name)
+
+        self.queue_service = None
+
+        # Track required event subscription updates
+        self.require_event_update = False
+
+        # Policy cache and dictionary
+        self.policy_cache = tempfile.mkdtemp()
```

Comment:
any reason why is the policy cache a temp directory?

Reply:
Yeah we don't use the disk cache heavily yet, and we could do this all in memory if we wanted, but I have some future enhancements planned to optimize container restarts that will use it more and then we'll make it (optionally) more persistent. 

Currently we only reference the disk cache when we see a YML file was removed or modified (hash changes).  We go back to the old copy we have on disk to see what policies used to be in it so we can clean up (unschedule jobs, remove event subscriptions) before we reconfigure the new policies.  It lets us take some shortcuts on updating things which means we can do it more often and reflect policy file changes very quickly.